ipfix/test_ipfix.py

```diff
@@ -0,0 +1,161 @@
+        self.logger.info(self.vapi.ppcli("ipfix flow-meter " + self.pg1.name))
+
+    def test_template(self):
+        self.logger.info(self.vapi.ppcli("set ipfix timeout template 1"))
+
+        self.pg0.enable_capture()
+        self.pg1.enable_capture()
+        self.pg_start()
+        # no packets on pg1 since we didn't inject any packets
+        self.pg1.assert_nothing_captured()
+        # expect 3 template pkt, 1/sec
+        capture = self.pg0.get_capture(timeout = 3, expected_count = 3)
+        self.verify_template(self.pg0, self.pg1, capture)
+
+    def verify_template(self, collector_if, exporter_if, capture):
+        for packet in capture:
+            try:
+                ip = packet[IP]
+                udp = packet[UDP]
+                ipfix = packet[IPFIX]
+                self.assert_equal(ip.src, exporter_if.remote_ip4,
+                                  "exporter ip")
+                self.assert_equal(ip.dst, collector_if.remote_ip4,
+                                  "collector ip")
+                self.assert_equal(udp.dport, 4739)
+                self.assert_equal(ipfix.version, 10)
+                # check that there's an IPFIX template
+                self.assert_equal(ipfix[Template].templateID, 256)
+            except IndexError:
+                self.logger.error(ppp("Invalid packet:", packet))
+                raise
+            except AssertionError:
+                self.logger.error(ppp("Unexpected packet:", packet))
+                raise
+
+class TestIPFIXData(VppTestCase):
+    """ IPFIX data test case """
+
+    @classmethod
+    def setUpClass(cls):
+        super(TestIPFIXData, cls).setUpClass()
+        cls.create_pg_interfaces(range(2))
+        for i in cls.pg_interfaces:
+            i.admin_up()
+            i.config_ip4()
+            i.resolve_arp()
+
+    def setUp(self):
+        super(TestIPFIXData, self).setUp()
+        self.logger.info(self.vapi.ppcli("set ipfix ip collector " + self.pg0.remote_ip4))
+        self.logger.info(self.vapi.ppcli("set ipfix ip exporter " + self.pg1.remote_ip4))
+        self.logger.info(self.vapi.ppcli("ipfix flow-meter " + self.pg1.name))
+
+    def test_basic(self):
+        self.logger.info(self.vapi.ppcli("set ipfix timeout template 20"))
+        self.logger.info(self.vapi.ppcli("set ipfix timeout idle 3"))
+
+        packet_count = 10
+        packets = self.create_stream(self.pg0, self.pg1, packet_count)
+
+        self.pg0.add_stream(packets)
+        self.pg0.enable_capture()
+        self.pg1.enable_capture()
+        self.pg_start()
+
+        capture1 = self.pg1.get_capture(timeout = 2, expected_count = packet_count)
+        # with the timeout above, we expect to get just a data packet
+        capture0 = self.pg0.get_capture(timeout = 5, expected_count = 1)
+        self.verify_capture(self.pg0, self.pg1, capture0)
+
+    def verify_capture(self, collector_if, exporter_if, capture):
```

Comment:
I wonder if it makes sense if it's inheriting from a base class to have a method for verifying each packet and then in the `verify_template` method you can then do like

```python
def verify_template(self, collector_if, exporter_if, capture):
    for packet in capture:
        try:
            self.verify_packet(packet)
            self.assert_equal(packet[IP][UDP][IPFIX][Template].templateID, 256)
        except IndexError:
            # stuff here
        except AssertionError:
            # stuff here
```

Another option could be to have a common verification method which just takes a `template=False` argument, or of course what it is now is fine too :)

WYDT?